projects/packages/partner/src/class-partner-coupon.php

```diff
@@ -127,27 +137,125 @@ public function catch_coupon( $plugin_slug, $redirect_location ) {
 	/**
 	 * Purge partner coupon.
 	 *
-	 * We automatically purge partner coupons after a certain amount of time to prevent
-	 * us from unnecessarily promoting a product for months or years in the future.
+	 * We try to remotely check if a coupon looks valid. We also automatically purge
+	 * partner coupons after a certain amount of time to prevent unnecessary look-ups
+	 * and/or promoting a product for months or years in the future due to unknown
+	 * errors.
 	 */
-	public function purge_coupon() {
+	public function maybe_purge_coupon() {
+		// Only run coupon checks on Jetpack admin pages.
+		// The "admin-ui" package is responsible for registering the Jetpack admin
+		// page for all Jetpack plugins and has hardcoded the settings page to be
+		// "jetpack", so we shouldn't need to allow for dynamic/custom values.
+		// phpcs:ignore WordPress.Security.NonceVerification.Recommended
+		if ( ! isset( $_GET['page'] ) || ! 'jetpack' === $_GET['page'] ) {
+			return;
+		}
+
+		if ( $this->maybe_purge_coupon_by_added_date() ) {
+			return;
+		}
+
+		$last_check = Jetpack_Options::get_raw_option( self::$last_check_option, 0 );
+
+		// Limit checks to happen every minute instead of on all page requests.
+		if ( $last_check && time() > $last_check + MINUTE_IN_SECONDS ) {
+			return;
+		}
+
+		Jetpack_Options::update_raw_option( self::$last_check_option, time() );
+
+		$this->maybe_purge_coupon_by_availability_check();
+	}
+
+	/**
+	 * Purge coupon based on local added date.
+	 *
+	 * We automatically remove the coupon after a month to "self-heal" if
+	 * something in the claim process has broken with the site.
+	 *
+	 * @return bool Return whether we should skip further purge checks.
+	 */
+	protected function maybe_purge_coupon_by_added_date() {
 		$date = Jetpack_Options::get_option( self::$added_option, '' );
 
 		if ( empty( $date ) ) {
-			return;
+			return true;
 		}
 
 		$expire_date = strtotime( '+30 days', $date );
 		$today       = time();
 
 		if ( $today >= $expire_date ) {
-			Jetpack_Options::delete_option(
-				array(
-					self::$coupon_option,
-					self::$added_option,
-				)
-			);
+			$this->delete_coupon_data();
+
+			return true;
+		}
+
+		return false;
+	}
+
+	/**
+	 * Purge coupon based on availability check.
+	 *
+	 * @param string|null $connection_client The connection client to remotely look up availability.
+	 * @return bool Return whether we deleted coupon data.
+	 */
+	protected function maybe_purge_coupon_by_availability_check( $connection_client = null ) {
+		// Allow for client dependency injection to allow for mock client during unit testing.
+		if (
+			empty( $connection_client ) ||
+			! class_exists( $connection_client ) ||
+			! method_exists( $connection_client, 'wpcom_json_api_request_as_blog' )
+		) {
+			$connection_client = Connection_Client::class;
+		}
```

Comment:
I think it would be better if we inject a callable in the class constructor instead e.g.:
```php
<?php
/**
 * Callable that executes a blog-authenticated request.
 *
 * @var callable
 */
protected $request_as_blog;

/**
 * Get singleton instance of class.
 *
 * @return Partner_Coupon
 */
public static function get_instance() {
    if ( is_null( self::$instance ) ) {
        self::$instance = new Partner_Coupon( [ Connection_Client::class, 'wpcom_json_api_request_as_blog' ] );
    }

    return self::$instance;
}

/**
 * Constructor.
 *
 * @param callable $request_as_blog Callable that executes a blog-authenticated request.
 */
public function __construct( $request_as_blog ) {
    $this->request_as_blog = $request_as_blog;
}
```

And then in this method we'd run this:
```php
$response = $this->$request_as_blog(
    sprintf(
        '/sites/%d/jetpack-partner/coupon/v1/site/coupon?coupon_code=%s',
        $jetpack_id,
        $coupon['coupon_code']
    ),
    2,
    array( 'method' => 'GET' ),
    null,
    'wpcom'
);
```

This way:
- we're not coupled to `Connection_Client::wpcom_json_api_request_as_blog()`
- the method that makes the request doesn't have to do parameter validation
- if we ever need to make a request in another method we don't have to pass the parameter again
- tests would be simpler (e.g. we won't need the Connection_Client mock class.)

Reply:
It also just fits better with DI, so a great suggestion! 👍

Reply:
That's where the inspiration came from - yes 😁